gooey/gui/processor.py

```diff
@@ -64,11 +66,21 @@ def _forward_stdout(self, process):
         Reads the stdout of `process` and forwards lines and progress
         to any interested subscribers
         '''
+        _start_time = get_current_time()

```

Comment:
Arg, so unfortunately, it didn't occur to me the first time I looked at this, but there's kind of a core design issue with this which is a show stopper in its current implementation. 

It's dependent on the assumption that there will be steady regular output being written to stdout. However if you run your code using something which produces chunked / irregular output (due to network latency or just general slow work or whatever), the elapsed time ceases to give steady updates. 

You can simulate the problem with a slight modification to your [example](https://github.com/chriskiehl/GooeyExamples/pull/31). 

```python
def compute_range(length):
    for i in range(10):
        import time
        from random import randint
        time.sleep(randint(1,3))
        print(f"progress: {i}/{10}")
```

This will cause the elapse time output to hop randomly between 1-3 seconds rather than giving a smooth continuous output. 

So, for this to work in the general case, you'll need to break apart the production of a timing event from the publishing of activity related to stdout. 

One way to accomplish this would be:

1. Introduce a stand alone timing class which deals with publishing updates about the current run time.
    * this class will have an internal blob of data which houses the timing info (start, estimatedRemaining) 
    * it'll have a subscription on `PROGRESS_UPDATE` which it can use update its internal estimation 
2. This class will publish its timing info on some regular cadence via [wx.Timer](https://wxpython.org/Phoenix/docs/html/wx.Timer.html#wx.Timer) 

As a super-duper rough sketch, the idea would be something like... 


```python
class Timing(object):
    def __init__(self, parent):
        self.startTime = 0
        self.estimatedRemaining = None
        self.wxTimer = wx.Timer(parent)
        parent.Bind(wx.EVT_TIMER, self.publishTime, self.wxTimer)

        pub.subscribe(events.PROGRESS_UPDATE, self._updateEstimate)

    def _updateEstimate(self, *args, **kwargs):
        prog = kwargs.get('progress')
        # your computation here
        self.estimatedRemaining = 0

    def publishTime(self, *args, **kwargs):
        pub.send_message(
            events.TIME_UPDATED,
            start=self.startTime,
            current=time.time(),
            elapsed=time.time() - self.startTime,
            estimatedRemaining=self.estimatedRemaining)

    def start(self):
        self.startTime = time.time()
        self.estimatedRemaining = None
        self.wxTimer.Start(500)

    def stop(self):
        self.wxTimer.Stop()
```

This would allow timing events to be published regardless of whether or not there was activity coming over the hosted program's stdout.

Reply:
Finally coming back around to looking at this, I'll have a go at we'll see how far we get abstracting it away from progress

Reply:
Resolved 😄  This is now fixed, I'm updating the tests for it and then I'll squash it all and push

Reply:
Apologies on my end for the delay! Glad you're back at it! Could you bring this up to date with 1.0.5? I wanna give it a whirl locally but it currently conflicts with the current state of things.